### DISCUSSION TITLE
How to serialize Item in Minestom and deserialize in Bukkit

### DISCUSSION THREAD

**Discussion by vallihalli** [Minestom/Minestom#647] [Help] [answered]
How to serialize Item in Minestom and deserialize in Bukkit? Serialized Item is to be saved in and loaded from SQL Database. Thanks

**Comment by TheMode** [comment, marked as answer]
The best way would be serializing the item into NBT. As the format is unrelated to the server backend.

You can use `ItemStack#toItemNBT`, or build up your own that should include material/amount/meta nbt

**Reply by vallihalli** [reply]
Thank you for taking the time and answering. How would I go about deserializing the item on the bukkit end though?

**Reply by TheMode** [reply]
The format returned by `toItemNBT` comes from Mojang, notably used in [bundle](https://minecraft.fandom.com/wiki/Bundle#Item_data), and IIRC for chat hover.

There is surely a way to load items from NBT using bukkit but I do not have the code at hand. [this](https://bukkit.org/threads/how-to-get-item-nbt.479264/) should explain the basics

**Reply by vallihalli** [reply]
But how do I serialize nbt

**Reply by TheMode** [reply]
`NBT#toSNBT()`

**Comment by vallihalli** [comment]
I found out how to do it now:
in Minestom do 
```java
final NBTCompound nbt = itemStack.toItemNBT().modify(root -> root.setInt("DataVersion", PAPER_DATA_VERSION));
final ByteArrayOutputStream destination = new ByteArrayOutputStream();
try (final NBTWriter writer = new NBTWriter(destination, CompressedProcesser.GZIP)) {
    writer.writeNamed("", nbt);
} catch (IOException e) {
    e.printStackTrace();
}
return destination.toByteArray();
```
to serialize it and in paper just use ItemStack.deserializeBytes
